### DISCUSSION TITLE
Roadmap?

### DISCUSSION THREAD

**Discussion by siefkenj** [protz/thunderbird-conversations#1572] [General]
I know the current focus is getting conversations to a point where the quick-compose can be re-enabled. Is there a roadmap for beyond that?

**Comment by Standard8** [comment]
I think quick compose could possibly be re-enabled in a basic form now, but needs probably a bit of code reorganisation and then styling work to allow it to run as quick reply or as the separate compose. See #1513. It won't be as fully featured as the old one was, but it would be a start.

With regards to the roadmap, there's probably a long term and short term split.

**Long Term**

Long term I would like to integrate as much as possible into Thunderbird itself. It could be that we integrate the whole add-on and change Thunderbird's default display - that's something I believe they are interested in, but there's still a lot of work before we get there, and isn't guaranteed. 

If we didn't get there, then I think the most significant, and useful part, of the needed changes would be to change how the Thunderbird message views work. Currently there is an awkward split between single and multiple email views, as well as things like random back end code assuming it can alter the views whenever it likes.

I believe if we could tidy some of that up, we could get to a stage where the message pane is essentially its own view, and it is passed messages with respect to what to display. An add-on could then replace the whole message pane layout with a different page, but it'd be able to function in exactly the same way.

Of course, if we get to the stage where Thunderbird is displaying the multiple messages for us, then it might just be a matter of styling, but we'll see when we get nearer.

**Short Term**

For the shorter term there's a few things I would like to hit on, though I obviously haven't had much time recently. In no particular order, though Quick Reply should probably be at the top.

* [Quick Reply](https://github.com/thunderbird-conversations/thunderbird-conversations/projects/3)
  * It has been too long without this, we need to get at least the basic version back again, and we can then build up the functionality.
* Finish the work I started to get a [column API into Thunderbird itself](https://github.com/thunderbird-conversations/thunderbird-conversations/issues/1430), and make the between column use it.
* Setup assistant. There's several things I'd like to do here:
  * Drop the automatic switching to the unified folder view. Make it an option with a couple of pictures to explain.
  * Remove the "option" for enabling the global indexer - do it by default, but use extension set preferences (*) to do so
  * Review the other options - can we move things to extension set preferences, do we provide more pictures for users to look at and decide? etc etc.
  * Overall I'd like to completely rework the UI and make it simpler and less wordy, and clearer for the bits where we're telling the user what we're changing.
* For the APIs that make sense move them into Thunderbird itself.
  * At the moment this probably isn't many of our existing ones, since a lot of them are just providing interfaces into the core of Thunderbird, though there's probably a few.
* Move the functionality in `addon/content/modules/` into other areas - either background scripts, APIs, or stub.xhtml.
  * contacts.js could probably go in the background as it is doing caching and it'd be good to keep that across folder changes.
  * conversations.js, message.js might be more appropriate with the functionality being extracted and driving stub.xhtml direct. A lot of what these are doing now is simply collating the information for the UI to display.
  * Need also to figure out about possible integration with other add-ons, e.g. [DKIMVerifier](https://github.com/lieser/dkim_verifier/issues/203)
* [Dark Mode](https://github.com/thunderbird-conversations/thunderbird-conversations/projects/5).
  * We need it (though I think it'll be complicated for message display).
* Fix bugs, and try and keep it stable.
  * I think it is a lot better than it has been, but we still need to be careful not to introduce regressions as we're rewriting it. Hopefully more tests will help here as well.

(*) "Extension set preferences" - Firefox has [some APIs](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/browserSettings) where add-ons can request the value of, or set, a preference within the core of Firefox. The requests get stacked, so that if an add-on is uninstalled, the preference reverts to what the user had set it to before - this would also negate the need for the "undo setup assistant" functionality.

I'd like to use the equivalent of these for ensuring the global indexer is enabled, setting the folder view etc. Obviously we'd need to work on adding these to Thunderbird itself.

** Conclusion **

For both the short term and long term there's a lot of work to be done. I think getting quick reply and the assistant simplified would be very good for users. Dark mode would also be a bonus. However, we also need to keep up the work of reworking the code, simplifying it and preparing for the future - especially with moving as much into Thunderbird APIs as we can.

**Reply by Standard8** [reply]
I've just updated the previous comment with a couple of links to open projects/issues. Hopefully will get time to do some more soon.

**Comment by Standard8** [comment]
Expanding on the part of my roadmap comment for conversations.js/message.js of "Move the functionality in addon/content/modules/ into other areas".

I think there is mainly a triplet of existing modules to consider moving roughly together which currently provide the main hooks for the conversations loading and management functionality.

Before I get into the details of that triplet, a slight complication is the plugin handlers / hooks. We probably need to come up with a solution for those before we do too much of the other rework.

I've [previously chatted](https://github.com/lieser/dkim_verifier/issues/203) with the author of the DKIMVerifier about how we might make external "plugins" work, and that approach seemed reasonable.

I'm not sure what we want to do about enigmail/lightning - I think they are worth having separate modules for, but I'm not yet sure how they should function.

The glodaAttrProviders really need to be moved across to WebExtension APIs.

Back to the triplet:

* contact.js
  * I think this is probably best moved to something loaded from the background script, and contact details accessed via runtime messages between the background script & the stub page.
  * The background script will allow us to cache the contacts for the session (maybe with a cache limit) which avoids expensive look ups.
  * Caching across stub.xhtml loads also means we can better keep the colours for the contacts consistent for the session, even when switching folders. It probably doesn't matter too much but would be nice.
  * `browser.runtime.sendMessage` won't work for stub.xhtml at the moment, but I did find a way to simulate it the other day, so we should be able to get a simulation working reasonably easily, and that will help for when we can move across to a WebExtension page.
* message.js
  * This is mainly a list of constructs to extract message data from the backend and put it in a format for the react data store. I think we should look at putting most of that into something in the reducer mechanism.
  * There's also various things that need to move across to WebExtension APIs (e.g. show remote content handling, phishing).
  * exportAsHtml is also an interesting one as that is used for forwarding. Maybe we could actually build that from the react store data?
* conversation.js
  * Effectively the heart of conversations.
  * A new object is created each time we change message selection. Something to be aware of here is changing selection quickly can cause us to still be getting details of one message set when we already have started loading the next. Hence, whatever we change here needs to take account of it. stub.xhtml is kept alive when switching threads within the same folder, so that's something to be aware of as well.
  * Note that new object creation is also currently handled in the `summarizeThreadHandler` in `msgWindowAPI.js`
  * I think we could move this logic into stub.xhtml, it makes sense for it to be there, I'm not quite sure how easily we could get the message selection into it - e.g. I know there's some callbacks going on and I'm not sure when the selected messages would get updated in relation to that loading. We should be able to experiment on it though.
  * I suspect there's also a few things in here that still need to move to WebExtension APIs as well, e.g. the gloda queries.

Overall, now I've written that lot, the starting points might be:

* Start moving contact.js to a background module. That's fairly standalone. Can we also get message.js to only pass the basic contact information to the store, and have the actions look it up as necessary?
* Explore what to do about the plugins.
* Can we move some or most of the message data loading/processing into stub.xhtml? e.g. we just pass the message id into the store, and actions retrieve the necessary data.
* Investigate if we could drive "Conversations" objects from stub.xhtml directly.

I'm totally open to debate/ideas on this. Treat the above as an initial brain dump. I've thought a lot about it in overview, but I haven't looked at all the details, so there might be a few things that won't work right if we do it this way.

**Reply by yoshimo** [reply]
Enigmail is no more. The openpgp support has been merged into TB 78 itself.

**Reply by Standard8** [reply]
> Enigmail is no more. The openpgp support has been merged into TB 78 itself.

Enigmail was merged into Thunderbird, just like Lightning was before, it was renamed slightly but still essentially a merge. The enigmail/lightning plugins are both still required for the respective functionality to work.

We could probably rename them when we do the rework, but they're definitely needed.

**Reply by siefkenj** [reply]
> browser.runtime.sendMessage won't work for stub.xhtml at the moment, but I did find a way to simulate it the other  day, so we should be able to get a simulation working reasonably easily, and that will help for when we can move across to a WebExtension page.

Can you explain how you simulated this?

**Comment by siefkenj** [comment]
I know you generally don't like transpiling, but have you considered switching to Typescript? I converted one of my projects recently and it revealed a lot of subtle bugs.

**Reply by Standard8** [reply]
I think that might be something to look at in future, but at this stage I think I'd rather focus on the current rewrites/improvements rather than add something extra into the mix.
